soroban-env-host/src/host/metered_vector.rs

```diff
@@ -1,250 +1,298 @@
-use super::{MeteredClone, MeteredCmp};
-use crate::xdr::ScHostObjErrorCode;
+use soroban_env_common::{xdr::ScHostFnErrorCode, Compare};
+
+use super::MeteredClone;
 use crate::{
-    budget::{Budget, CostType},
-    HostError,
+    budget::{AsBudget, Budget, CostType},
+    xdr::ScHostObjErrorCode,
+    Host, HostError,
 };
-use im_rc::{vector::Iter, vector::IterMut, Vector};
-use std::cmp::{min, Ordering};
-use std::ops::RangeBounds;
-use std::rc::Rc;
-
-pub struct MeteredVector<A> {
-    budget: Budget,
-    vec: Vector<A>,
+use std::{cmp::Ordering, ops::Range};
+
+#[derive(Clone)]
+pub struct MeteredVector<A>
+where
+    A: MeteredClone,
+{
+    vec: Vec<A>,
 }
 
-impl<A> MeteredVector<A> {
-    fn charge_new(&self) -> Result<(), HostError> {
-        self.budget.charge(CostType::ImVecNew, 1)
+impl<A> MeteredVector<A>
+where
+    A: MeteredClone,
+{
+    fn charge_new(&self, size: usize, budget: &Budget) -> Result<(), HostError> {
+        budget.charge(CostType::VecNew, size as u64)
     }
 
-    fn charge_mut_access(&self, x: u64) -> Result<(), HostError> {
-        self.budget.charge(CostType::ImVecMutEntry, x)
+    fn charge_access(&self, count: usize, budget: &Budget) -> Result<(), HostError> {
+        budget.charge(CostType::VecEntry, count as u64)
```

Comment:
Are the `CostType::VecEntry` and `VecNew` charges are dependent on the element type `A`?
- For `VecNew` I think so since it needs to pre-allocate memory depending on `size_of<A>`.
- For `VecEntry` I think not because it's just moving a pointer? 

Am I correct?

Reply:
I think so, but as you mentioned above `size_of` is not stable / we can't rely on it. So we should just capture the linear input size and treat all such costs as the same.